_posts/2020-09-21-centralised-schema-management-microservices.md

```diff
@@ -0,0 +1,209 @@
+   - At a minimum, your consumer must be able to tell you __if a data structure is valid or not__ in the context of a schema.
+   - If it can also do __coercion__ of _almost_ valid data into _completely_ valid data, like parsing an ISO8601 string into a DateTime object, that's good.
+- Schema distribution to consumers can be done during compilation or packaging, but it's more microservicey if you have a __schema registry service__. It can just be nginx serving static files, in a pinch.
+
+## We're not in MVP land anymore
+
+The above is fine for a workshop demo. However, it's likely your system will need something a bit more advanced to survive in the wild. To upgrade your CSM from the "Home Edition" into "Professional", you'll probably need these:
+
+- __Support for schema evolution__. Schemas are like war plans, they never survive enemy contact. You need to prepare for this by having a way to handle changes to your schemas without the need to rebuild everything and breaking all your external APIs.
+  - But what to do if you end up with an object of version 1, and you need to upgrade it into version 3? Or the reverse?
+  - __Transform__ is what you do. Your CSM system can help by providing transformation services, or supplying transformation instructions alongside your schemas. To do this efficiently and in a way compatible with as many languages and platforms as possible is difficult, though. More on this later on.
+- __Support for schema composition__. If you need CSM to begin with, then your data types probably have things in common between them. You'll want to be able to compose your schemas from components. It's boring and error-prone to copy-paste the same list of `create-time`, `created-by`, `modify-time` and `modified-by` properties to every schema.
+  - I don't need to mention this complicates evolution and transformation somewhat, do I?
+  - You could also consider deriving schemas from other schemas, but that smells like OOP-style inheritance and we don't do that.
+- __Enumerations as first-class citizens__. Real-world systems always have various lists of enumeration-type values that evolve over time. You need to be able to apply the same tools to them as the rest of your schemas: evolution, transformation, etc.
+- __API management integration__ &mdash; if you can connect your CSM with your API management solution you can have it handle inbound/outbound data validation for you.
+
+## Nah, man. I'm pretty friggin' far from MVP
+
+In for a penny, in for a pound? Why not go for the whole hog? Since you're already committed, why leave money on the table by not extracting some more value from your fancy bespoke CSM? To evolve your CSM into its final form you could...
+
+- __Embed business logic into your schemas.__ If your CSM supports arbitrary metadata you can include things such as UX rendering instructions alongside your property definitions.
```

Comment:
Maybe you could give a more concrete example of how this would be a good thing? Your example of UX rendering feels like an anti-pattern to me. How is "embedded elastic search mapping" business logic and not just a declarative mapping? Is "trivial transformations between different schema versions" the only sensible use case? Unless I'm just being a bit too sceptical and would like to see something more concrete.